rfcs/QueryLevelNullability.md

```diff
@@ -0,0 +1,179 @@
+error to the client rather than any data.
+
+On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.
+On mobile platforms where codegen is used, clients have full control over the nullability of the properties on the
+generated types. Since nullability is expressed in the query rather than the schema, it's flexible enough to accommodate
+various use-cases (e.g., where the business `name` _is_ allowed to be nullable).
+
+### `!`
+
+We have chosen `!` because `!` is already being used in the GraphQL spec to indicate that a field is non-nullable.
+Incidentally the same precedent exists in Swift (`!`) and Kotlin (`!!`) which both use similar syntax to indicate
+"throw an exception if this value is `null`". 
+
+### Use cases
+
+#### When a field is necessary to the function of the client
+There are times where a field can be `null`, but a feature that queries for the field will not function
+if it is `null`. For example if you are trying to render an information page for a movie, you won't
+be able to do that if the name field of the movie is missing. In that case it would be preferable
+to fail as early as possible.
+
+### ✨ Examples
+
+#### Non-nullable field
+```graphql
+query GetBusinessName($encid: String!) {
+  business(encid: $encid) {
+    name!
+  }
+}
+```
+would codegen to the following type on iOS.
+
+```swift
+struct GetBusinessNameQuery.Data.Business {
+  let name: String // lack of `?` indicates that `name` will never be `null`
+}
+```
+
+#### Non-nullable object with nullable fields
+```graphql
+query {
+  business(id: 4) {
+    reviews! {
+      rating
+      text
+    }
+  }
+}
+```
+
+## ✅ RFC Goals
+- Non-nullable syntax that is based off of syntax that developers will already be familiar with
+
+## 🚫 RFC Non-goals
+
+## 🗳️ Alternatives considered
+
+### Make Schema Fields Non-Nullable Instead
+Discussion on [this topic can be found here](https://medium.com/@calebmer/when-to-use-graphql-non-null-fields-4059337f6fc8)
+// We definitely want to flesh this out
+
+### Use a directive such as `@non-null` instead
+This is the alternative being used at some of the companies represented in this proposal for the time being.
+However this feels like a common enough need to call for a language feature. A single language feature
+rather than using directives as a workaround would enable more unified public tooling around GraphQL.
+
+### Write wrapper types that null-check fields
+This is the alternative being used at some of the companies represented in this proposal for the time being.
+It's quite labor intensive and the work is quite rote. It more or less undermines the purpose of
+having code generation.
+
+### Alternatives to `!`
+#### `!!`
+This would follow the precedent set by Kotlin.
+
+### Make non-nullability apply recursively
+For example, everything in this tree would be non-nullable
+```graphql
+query! {
+  business(id: 4) {
+    name
+  }
+}
+```
```

Comment:
@wxue Do you remember why we didn't want this?

Reply:
I think we should remove this as an alternative, I don't think we understand it well enough to vouch for it